wasm/src/main/java/com/dylibso/chicory/wasm/WasmModule.java

```diff
@@ -0,0 +1,294 @@
+    }
+
+    public ElementSection elementSection() {
+        return elementSection;
+    }
+
+    public void setElementSection(ElementSection elementSection) {
+        this.elementSection = requireNonNull(elementSection);
+    }
+
+    /**
+     * Creates a {@link Builder} for the specified {@link InputStream}
+     *
+     * @param input the input stream
+     * @return a {@link Builder} for reading the module definition from the specified input stream
+     */
+    public static Builder builder(InputStream input) {
+        return new Builder(() -> input);
+    }
+
+    /**
+     * Creates a {@link Builder} for the specified {@link WasmModule}
+     *
+     * @param wasmModule the already parsed Wasm module
+     * @return a {@link Builder} for reading the module definition from the specified input stream
+     */
+    public static Builder builder(WasmModule wasmModule) {
+        return new Builder(wasmModule);
+    }
+
+    /**
+     * Creates a {@link Builder} for the specified {@link ByteBuffer}
+     *
+     * @param buffer the buffer
+     * @return a {@link Builder} for reading the module definition from the specified buffer
+     */
+    public static Builder builder(ByteBuffer buffer) {
+        return builder(buffer.array());
+    }
+
+    /**
+     * Creates a {@link Builder} for the specified byte array
+     *
+     * @param buffer the buffer
+     * @return a {@link Builder} for reading the module definition from the specified buffer
+     */
+    public static Builder builder(byte[] buffer) {
+        return new Builder(() -> new ByteArrayInputStream(buffer));
+    }
+
+    /**
+     * Creates a {@link Builder} for the specified {@link File} resource
+     *
+     * @param file the path of the resource
+     * @return a {@link Builder} for reading the module definition from the specified file
+     */
+    public static Builder builder(File file) {
+        return new Builder(
+                () -> {
+                    try {
+                        return new FileInputStream(file);
+                    } catch (FileNotFoundException e) {
+                        throw new IllegalArgumentException(
+                                "File not found at path: " + file.getPath(), e);
+                    }
+                });
+    }
+
+    /**
+     * Creates a {@link Builder} for the specified {@link Path} resource
+     *
+     * @param path the path of the resource
+     * @return a {@link Builder} for reading the module definition from the specified path
+     */
+    public static Builder builder(Path path) {
+        return new Builder(
+                () -> {
+                    try {
+                        return Files.newInputStream(path);
+                    } catch (IOException e) {
+                        throw new IllegalArgumentException("Error opening file: " + path, e);
+                    }
+                });
+    }
+
+    public static class Builder {
```

Comment:
I dislike having a builder just to read a file. The builder only has two options: logger and file format.

The logger is only used in one place, to warn about ignored section IDs. If anyone cares about that, we could collect them and have a method `getIgnoredSectionIds()`. Or perhaps we should fail, since an unknown ID means the file is malformed or non-compliant. We can see what other implementations do. Thus, we can remove the logger.

The file format is not used today. It's also unclear if we would want to load the text format using the same API. The browser API only supports the binary format. Using `Wat2Wasm` or similar to covert to binary first seems like a reasonable answer. Another future option is to add auto-detection: try to parse as text if the file does not start with the binary magic value.

With those changes, we can have a simpler API:
```java
WasmModule module = WasmModule.load(path);
```
All of the `builder()` methods become `load()` methods that return `WasmModule`.

Reply:
I do understand and love this comment, I'm a little worried of applying it straight away(i.e. in this PR) as it will change a lot again ...
Let me go through the rest of the comments and see how far we can get.

Reply:
I ended up doing those changes, I'll stick with the `builder` since handling `InputStream`s, in memory `byte[]` etc. seems to be compelling use cases, but simplifying where possible and removing redundant APIs.

Reply:
We can handle all of these with overloaded `load()` methods. I wasn't suggesting to only support files. But it doesn't require a builder. I can follow up on this if you don't want to change it here.

Reply:
I started doing those changes, but I would like to have the discussion about the support for text format etc.
I think it is better to do it in its own PR.

I have some mixed feelings here:
- having a single `load` method is extremely convenient
- a single method makes the code less "future-proof"
- adding overloads to the "load" method becomes maintenance burden

The reason I'm heading to this direction is that, for example, Component Model changes the Magic Number:
https://github.com/WebAssembly/component-model/blob/main/design/mvp/Binary.md#component-definitions

I don't know if/when/how we would support CM(and/or other features), but it seems appealing to leave the possibility to add properties to a `Builder` even at the `Parser` level, I think it will become handy when implementing extensions to the spec.